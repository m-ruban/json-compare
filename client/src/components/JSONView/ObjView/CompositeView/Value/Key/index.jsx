import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Mark from 'components/Mark';

const Key = ({ value, isArray }) => {
    if (isArray) {
        return <Mark type="number" text={value} />;
    }

    return <Fragment>&quot;{value}&quot;</Fragment>;
};

Key.propTypes = {
    value: PropTypes.string,
    isArray: PropTypes.bool,
};

export default Key;
