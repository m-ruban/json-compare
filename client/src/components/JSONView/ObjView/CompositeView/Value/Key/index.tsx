import React from 'react';

import Mark from 'components/Mark';

const Key = ({ value, isArray }: { value: string; isArray: boolean; }) => {
    if (isArray) {
        return <Mark type="number" text={value} />;
    }
    return <span>{'"'}{value}{'"'}</span>;
};

export default Key;
