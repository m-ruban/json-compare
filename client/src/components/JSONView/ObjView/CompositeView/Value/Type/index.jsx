import React from 'react';
import PropTypes from 'prop-types';
import { getType } from 'modules/compare';
import Mark from 'components/Mark';

const Type = ({ value }) => {
    const type = getType(value);
    return <Mark type={type} text={type} />;
};

Type.propTypes = {
    value: PropTypes.any,
};

export default Type;
