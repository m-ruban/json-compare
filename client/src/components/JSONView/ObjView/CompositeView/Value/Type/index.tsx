import React from 'react';

import Mark from 'components/Mark';
import { getType } from 'modules/compare';

const Type = ({ value }: { value: string; }) => {
    const type: string = getType(value);
    return <Mark type={type} text={type} />;
};

export default Type;
