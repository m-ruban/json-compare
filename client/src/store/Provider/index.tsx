import * as React from 'react';

import { COMPARE_REQ } from 'constants/compare';
import { EQUALITY_TYPE_STRICT } from 'constants/equality';
import { AUTHOR_LINK, FEEDBACK_LINK } from 'constants/links';
import { LOG_INFO } from 'constants/log';
import { Provider } from 'store/Context';

import { equalAlerts } from 'modules/compare';
import { parseAndLog } from 'modules/parse';

interface IProviderProps {
    children: any;
}

interface IProviderState {
    alerts: Map<string, object>;
    author: string;
    changeEqualityTypes: (event: React.SyntheticEvent) => void;
    changeObj: (event: React.SyntheticEvent) => void;
    changeTool: (value: string) => void;
    equalityTypes: string;
    feedback: string;
    isCompare: boolean;
    log: object;
    objObject: object;
    objStr: string;
    toggleCompare: (event: React.SyntheticEvent) => void;
    toolObject: object;
    toolStr: string;
}

class AppProvider extends React.Component<IProviderProps, IProviderState> {
    constructor(props) {
        super(props);
        this.state = {
            alerts: new Map(),
            author: AUTHOR_LINK,
            changeEqualityTypes: this.changeEqualityTypes.bind(this),
            changeObj: this.changeObj.bind(this),
            changeTool: this.changeTool.bind(this),
            equalityTypes: EQUALITY_TYPE_STRICT,
            feedback: FEEDBACK_LINK,
            isCompare: false,
            log: { text: '', type: null },
            objObject: {},
            objStr: '',
            toggleCompare: this.toggleCompare.bind(this),
            toolObject: {},
            toolStr: '',
        };
    }

    public toggleCompare() {
        const log = {
            text: 'toggle compare',
            type: LOG_INFO,
        };

        this.setState({ isCompare: !this.state.isCompare, log }, this.changeAlerts);
    }

    public changeEqualityTypes(event) {
        const { value } = event.target;
        const log = {
            text: 'change equality types',
            type: LOG_INFO,
        };
        this.setState({ equalityTypes: value, log }, this.changeAlerts);
    }

    public changeTool(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ toolObject: obj, toolStr: value, log }, this.changeAlerts);
    }

    public changeObj(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ objObject: obj, objStr: value, log }, this.changeAlerts);
    }

    public changeAlerts() {
        const { toolObject, objObject, equalityTypes, isCompare } = this.state;
        const isEmpty = (obj) => Object.keys(obj).length === 0;

        if (!isEmpty(toolObject) && !isEmpty(objObject) && isCompare) {
            const alerts = equalAlerts(toolObject, objObject, equalityTypes);
            let req = 0;
            let diff = 0;
            let text = '';

            alerts.forEach((alert) => {
                if (!alert.scalar) { return; }
                if (alert.res === COMPARE_REQ) {
                    req++;
                } else {
                    diff++;
                }
            });

            if (diff > 0 || req > 0) {
                text = `${diff} alert${diff > 1 ? 's' : ''}, ${req} required`;
            } else {
                text = 'objects are equal';
            }

            this.setState({
                alerts,
                log: { text, type: LOG_INFO },
            });
        }
    }

    public render() {
        return <Provider value={this.state}>{this.props.children}</Provider>;
    }
}

export default AppProvider;
