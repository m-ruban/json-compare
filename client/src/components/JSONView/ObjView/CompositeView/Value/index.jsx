import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import ObjView from 'components/JSONView/ObjView';
import Key from 'components/JSONView/ObjView/CompositeView/Value/Key';
import Type from 'components/JSONView/ObjView/CompositeView/Value/Type';

const Value = ({ obj, lvl, path, showAlerts, showRequired, isArray }) => (
    <Fragment>
        {Object.entries(obj).map(([key, value]) => {
            const newPath = path.slice();
            const margin = `${lvl * 6}px`;
            const divKey = `${key}_${lvl}`;
            newPath.push(key);

            return (
                <div key={divKey} style={{ marginLeft: margin }}>
                    <Key value={key} isArray={isArray} />
                    :&nbsp;
                    <Type value={value} />
                    <ObjView
                        value={value}
                        path={newPath}
                        lvl={lvl}
                        showAlerts={showAlerts}
                        showRequired={showRequired}
                    />
                </div>
            );
        })}
    </Fragment>
);

Value.propTypes = {
    obj: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
    lvl: PropTypes.number,
    path: PropTypes.array,
    showAlerts: PropTypes.bool,
    showRequired: PropTypes.bool,
    isArray: PropTypes.bool,
};

export default Value;
