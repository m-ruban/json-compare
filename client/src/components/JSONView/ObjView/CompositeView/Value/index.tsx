import React from 'react';

import ObjView from 'components/JSONView/ObjView';
import Key from 'components/JSONView/ObjView/CompositeView/Value/Key';
import Type from 'components/JSONView/ObjView/CompositeView/Value/Type';

interface IValue {
    obj: any[] | object;
    lvl: number;
    path: string[];
    showAlerts: boolean;
    showRequired: boolean;
    isArray: boolean;
}

const Value = ({ obj, lvl, path, showAlerts, showRequired, isArray }: IValue) => (
    <>
        {Object.entries(obj).map(([key, value]) => {
            const newPath = path.slice();
            const margin = `${lvl * 6}px`;
            const divKey = `${key}_${lvl}`;
            newPath.push(key);

            return (
                <div key={divKey} style={{ marginLeft: margin }}>
                    <Key value={key} isArray={isArray} />
                    :&nbsp;
                    <Type value={value} />
                    <ObjView
                        value={value}
                        path={newPath}
                        lvl={lvl}
                        showAlerts={showAlerts}
                        showRequired={showRequired}
                    />
                </div>
            );
        })}
    </>
);

export default Value;
