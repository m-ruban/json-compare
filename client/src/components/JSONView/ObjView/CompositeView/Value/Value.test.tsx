import React from 'react';

import Value from 'components/JSONView/ObjView/CompositeView/Value';
import { render } from 'enzyme';
import { Provider } from 'store/Context';
import alertsExample from 'tests/mocks/alerts';

describe('<Value />', () => {
    it('Renders correctly', () => {
        const props = { obj: {}, lvl: 1, path: [], showAlerts: false, showRequired: false, isArray: false };
        const output = render(<Value {...props} />);
        expect(output).toMatchSnapshot();
    });
    it('Renders correctly (obj)', () => {
        const props = { obj: {id: 1}, lvl: 1, path: [], showAlerts: false, showRequired: false, isArray: false };
        const output = render(
            <Provider value={alertsExample}>
                <Value {...props} />
            </Provider>,
        );
        expect(output).toMatchSnapshot();
    });
});
