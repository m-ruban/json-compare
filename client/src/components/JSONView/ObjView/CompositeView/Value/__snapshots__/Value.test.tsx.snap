// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Value /> Renders correctly (obj) 1`] = `
<div
  style="margin-left:6px"
>
  <span>
    "id"
  </span>
  :Â 
  <span
    class="json-mark json-mark_number badge badge-json"
  >
    number
  </span>
  <span>
    1
  </span>
</div>
`;

exports[`<Value /> Renders correctly 1`] = `null`;
